Analysis local storage

schlawg opened the pull request:
I wrote about this in discord.  It is very cool (for correspondence users at least) and I had hoped to get someone to look at it.

ornicar commented:
A nice feature in a maintainable amount of code.

One thing I'm concerned about is that people may think the analysis state is persisted on the server, like for studies. They might expect it to be available to other devices and other players, and to survive browser cache clearing.

schlawg commented:
Yes that's definitely a problem.  So it looks like we are definitely going to have to add a background sync mechanism between the client and lila.  haha relax I'm joking.

It probably needs to be an opt-out feature where turning it off deletes the database.  Could it go in the move list hamburger?

I would like to overhaul preferences a bit and move username/password/2fa/tokens/close account to an Account sub menu, partly because they're linked but also to make room for other categories.  Analysis pane because some move list hamburger options don't vary game to game and should be explicitly configurable as defaults.  This could free up needed space in the hamburger, or it could just mirror/parameterize some of those options.  Notifications pane is another I'd like to add.

ornicar commented:
I don't see how the opt-out and the preference overhaul help with the problem I mentioned. I'll try to display a warning message on first usage instead.

schlawg commented:
I was thinking text when you mouse over the switch.  "Save local moves in browser cache"

schlawg commented:
<img width="386" alt="Screen Shot 2022-07-22 at 2 37 58 PM" src="https://user-images.githubusercontent.com/101470903/180515596-d98de419-e6dd-44ec-9906-92a7ab6417f9.png">
<img width="355" alt="Screen Shot 2022-07-22 at 2 39 10 PM" src="https://user-images.githubusercontent.com/101470903/180515627-f2d70a19-43f7-4ab0-b1c2-485c6bbeecf0.png">

Yes that mouseover text is terrible but you get the idea.

ornicar commented:
The thing with settings is that people rarely use them. You can assume everyone will ignore them completely, stick to the defaults, and never read the explanations.

Maybe whenever custom moves are in the tree, we could display a pane at the bottom of the move list. It would say that moves are temporarily saved, that you can make a study for permanent storage. It would have an option to clear the current board.
Hell, maybe later we'll need to have it link to a page that lists all saved analysis, but we're not there yet.

Thoughts?

schlawg commented:
That's cool.  Since most would rather use that space for additional moves, I think you still need a switch to turn the feature on/off and reveal/hide the pane.  Does it go in the hamburger as some permutation of "Save moves in cache"?

BTW, there IS a clear current board button (reset icon) next to the hamburger but I agree it would be better as text in a pane.

ornicar commented:
?

schlawg commented:
I'm happy with this.

Bang on it when you get a chance and let me know what to change.  REC button doesn't behave exactly like study..  When you turn it ON, it saves the current moves to db rather than rewinding you to the state when it was last on.  I played with this for an hour or so and it's pretty intuitive.  The previous version where it was a setting that blows away your db when you turn it off was terrible.

schlawg commented:
Practice mode used to show a tiny pane when on before I had a minor git catastrophe and started merging stuff for the hell of it.  There is no practice view now.  Did I break deps somewhere?

```
                      retroView(ctrl) || deps?.practiceView(ctrl) || explorerView(ctrl) || renderSaveMoves(ctrl),
```

ornicar commented:
your turn to review code and test the UI :dancers:

schlawg commented:
Your changes are awesome.  I like it.

I'm worried they will think they need to have the button active (and the pane visible) in order to record their moves.  Popping it up for every new game or study would reinforce that so I changed that.

I also hide the button entirely when REC mode is active in study mode.

There were other fixes that had to do with multiple green buttons being active at the same time.  I changed some things for my own sanity, and hopefully someone familiar with the history of the toggle stuff in ctrl.ts will fix it all and join me for mead in Valhalla.

ornicar commented:
I certainly don't want that feature in a study. It's way too confusing to have both local AND server storage in there.

ornicar commented:
I'm going to merge up to my last commit.

> I'm worried they will think they need to have the button active (and the pane visible) in order to record their moves

Indeed. I'll see about adding a visual clue about it.

ornicar commented:
merged manually up to [92cdafe](https://github.com/lichess-org/lila/pull/11234/commits/92cdafe1c6a25ba6ba6807b1b1fc4ca6301ea2c0)
